orchestrator.go

```diff
@@ -0,0 +1,282 @@
+			duration := time.Since(t0)
+
+			if err != nil {
+				// Record the error but continue on no matter what, as empty sites will propogate soft shutdowns automatically.
+				// Watchers will be closed below, and the event queue will empty out itself.
+				logger.Error("getSites: could not get sites list; err: %v", err)
+				orch.metrics.RecordGetSites(false, duration)
+			} else {
+				orch.metrics.RecordGetSites(true, duration)
+			}
+
+			sitesToWatch, sitesToClose := checkForSiteDiffs(watchedSites, fetchedSites)
+
+			// Start up individual routines per each site we're watching.
+			for _, site := range sitesToWatch {
+				logger.Debug("starting thread %q", site.URL)
+				closeChan := make(chan struct{})
+				watchedSites[site.URL] = closeChan
+				go orch.startSiteWatcher(site, closeChan)
+			}
+
+			// If the site no longer exists, send the close signal and stop tracking.
+			for _, siteURL := range sitesToClose {
+				logger.Debug("closing thread %q", siteURL)
+				close(watchedSites[siteURL])
+				delete(watchedSites, siteURL)
+			}
+		}
+	}
+
+	return nil
+}
+
+/*
+ * Site watchers poll for new events every "getEvents" interval, and push them into the events queue.
+ *
+ * These are gated by a semaphore as a means of rate limiting. This helps ensure that:
+ * A) we don't spam the DB too hard all at once
+ * B) the events that were fetched won't grow super stale while waiting to be pushed into the events queue
+ */
+func (orch *Orchestrator) startSiteWatcher(site Site, close chan struct{}) {
+	// TODO splay out initial ticks
+	ticker := time.NewTicker(orch.config.GetEventsInterval)
+	defer ticker.Stop()
+
+	for {
+		select {
+		case <-close:
+			return // closed while waiting for ticker
+		case <-ticker.C:
+			select {
+			case <-close:
+				return // closed while waiting for semaphore
+			case orch.semGetEvents <- true:
+				// we have acquired the semaphore
+				orch.fetchSiteEvents(site, close)
+			}
+		}
+	}
+}
+
+func (orch *Orchestrator) fetchSiteEvents(site Site, close chan struct{}) {
+	defer (func() {
+		<-orch.semGetEvents
+	})()
+
+	t0 := time.Now()
+	events, err := orch.performer.GetEvents(site)
+	if err != nil {
+		logger.Error("getEvents: could not get events for site %s; err=%v", site.URL, err)
+	} else {
+		if len(events) > 20 {
+			logger.Warning("getEvents: got a lot of events (%d) for site %s", len(events), site.URL)
+		} else {
+			logger.Debug("getEvents: got %d events for site %s", len(events), site.URL)
+		}
+	}
+
+	orch.metrics.RecordGetSiteEvents(err == nil, time.Since(t0), site.URL, len(events))
+
+	for _, event := range events {
+		select {
+		case <-close:
+			return // closed while waiting for space in event queue
+		case orch.events <- event:
+			continue // noop, event is sent!
+		}
+	}
+}
+
+/*
+ * Runners are much more simple. If there is an event in the queue, run it!
+ */
+func (orch *Orchestrator) setupRunners() error {
+	activeRunners := make(threadTracker)
+	var wg sync.WaitGroup
```

Comment:
```suggestion
	wg := &sync.WaitGroup{}
	wg.Add(orch.config.NumRunWorkers)
```

you want to add all them at the same time, iirc it is a runtime panic to call `.Add` after anybody has already called `.Done` and although unlikely, we do not want it to happen.